### DISCUSSION TITLE
Script-defined compiler options

### DISCUSSION THREAD

**Discussion by vanillajonathan** [PRQL/prql#3481] [Ideas]
What if the script could define compiler options?

Example:
```elm
prql.signature_comment = false
prql.sql.format_style = "river"
prql.sql.lowercase = true
prql.sql.use_brackets = true

from tracks
select {artist, title}
sort title
```

Output:
```sql
  select [artist], [title]
    from [tracks]
order by [title]
```

**Comment by max-sixty** [comment]
Yes, we could offer more options in the header, like `signature_comment`: https://prql-lang.org/book/project/target.html

**Reply by eitsupi** [reply]
This reminds me, it would be great to be able to write prql header on multiple lines.

**Reply by max-sixty** [reply]
Do you think the new line wrapping is sufficient?

**Reply by eitsupi** [reply]
> Do you think the new line wrapping is sufficient?

I think so. Does this already work?

**Reply by max-sixty** [reply]
I just tried and I think so! This correctly fails with a header specifying a newer version than the current compiler:

```sh
echo 'prql
   \ version:"0.10.0"

   from tracks' | cargo run -p prqlc -- compile
   
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/prqlc compile`
Error: This query uses a version of PRQL that is not supported by the prql-compiler. Please upgrade the compiler.

```

**Reply by eitsupi** [reply]
That's great! Thanks for trying.

**Comment by snth** [comment]
I think this is a good idea. I am in support of it to the extent of exposing any options that `sqlformat` provides. I don't think we should be spending our time formatting SQL code since that's for machine consumption (since we're all writing PRQL right 😉).

---
I hadn't heard of the "River" formatting style for SQL before but found this: [Modern SQL Style Guide](https://gist.github.com/mattmc3/38a85e6a4ca1093816c08d4815fbebfb)

I like it! Thanks @vanillajonathan .

I would like to explore supporting River formatting for PRQL code!

**Reply by snth** [reply]
I believe we currently use the `sqlformat` crate for SQL formatting so the available options would be found here: https://docs.rs/sqlformat/latest/sqlformat/struct.FormatOptions.html

**Reply by snth** [reply]
That [Modern SQL Style Guide](https://gist.github.com/mattmc3/38a85e6a4ca1093816c08d4815fbebfb) also has an interesting bit on bracket-escaping table and column names. It's something I've been thinking about for a while and I'll write up in a separate discussion.

**Reply by max-sixty** [reply]
(I do think our options are somewhat limited on SQL formatting specifically, so discussing various formatting options has fairly low returns unless there's some willingness to make a fairly large contribution upstream)

**Reply by vanillajonathan** [reply]
I came across the "river style" formatting in the [SQL Style Guide](https://www.sqlstyle.guide/) by Simon Holywell.
